### PR TITLE
Add benchmarking CI that generates graphs on updates to master

### PR DESCRIPTION
**Pull request by jprider63**
This PR adds a CI workflow that generates performance graphs on every update to master. It stores the data as json in the `gh-pages` branch and renders the graphs on a Github Page. Here's [an example](https://galoisinc.github.io/cerberus/dev/bench/) of what this looks like with [dummy data](https://github.com/GaloisInc/cerberus/blob/gh-pages/dev/bench/data.js). 

This currently runs `cn` on all the (successful) .c files in the test suite. Eventually, we probably want to make a proper benchmark suite using something like [Core_bench](https://blog.janestreet.com/core_bench-micro-benchmarking-for-ocaml/).

**Before this is merged**, someone with permission needs to create a branch called `gh-pages` with empty contents. You can do this by grabbing the branch [`gh-pages-init`](https://github.com/GaloisInc/cerberus/tree/gh-pages-init) from our fork and pushing to `gh-pages` in this repository. 

Implements part of rems-project/cn-tutorial/issues/54.

### CODE REVIEW

File: tests/run-ci-benchmarks.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]: Usually we start with
```bash
#!/usr/bin/env bash
set -euo pipefail -o noclobber
# set -xv # uncomment to debug variables
```

File: tests/run-ci-benchmarks.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+JSON_FILE="benchmark-data.json"
+
+cat << EOF >> ${JSON_FILE}
+[
+EOF
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]:
```bash
echo "[" >> "${JSON_FILE}"
```

File: tests/run-ci-benchmarks.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+DIRNAME=$(dirname "$0")
+
+SUCC=$(find "${DIRNAME}"/cn -name '*.c' | grep -v '\.error\.c')
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]:
```bash
SUCC=($(find "${DIRNAME}"/cn -name '*.c' | grep -v '\.error\.c'))
```
But measuring the time taken for failing benchmarks would be useful too right?

File: tests/run-ci-benchmarks.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+COUNT=0
+for TEST in ${SUCC}; do
+  let COUNT=${COUNT}+1
+done
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]:
```bash
LEN=${#SUCC[@]}
```

File: tests/run-ci-benchmarks.sh [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+INDEX=0
+echo $SUCC
+echo $COUNT
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]: Is echoing the whole array necessary?

File: tests/run-ci-benchmarks.sh [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+for TEST in ${SUCC}; do
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]: You can iterate through the keys: http://stackoverflow.com/questions/3112687/ddg#3113285 so INDEX is unnecessary.

File: tests/run-ci-benchmarks.sh [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+EOF
+  else
+    cat << EOF >> ${JSON_FILE}
+    {
+	"name": "${TEST}",
+        "unit": "Seconds",
+	"value": ${TIME}
+    },
+EOF
+  fi
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh:43]: Perhaps printf and a variable for the last comma would reduce code duplication here.

File: tests/run-ci-benchmarks.sh [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+  let INDEX=${INDEX}+1
+done
+
+cat << EOF >> ${JSON_FILE}
+]
+EOF
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]:
```bash
echo "]" >> "${JSON_FILE}"
```

File: tests/run-ci-benchmarks.sh [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+cat ${JSON_FILE}
```
Comment (dc-mak) [inline tests/run-ci-benchmarks.sh]: Optional tip: (Installing and) using `jq` will colorise (and format) the output.

**Review by dc-mak — COMMENTED** [review]
If we envision this being updated and getting more complicated, then it would be worth considering a Python or OCaml script, but this is quite useful to have sooner than later so I'm willing to merge now and get the highlighted changes later.

Also consider using https://www.shellcheck.net/

File: tests/run-ci-benchmarks.sh (hunk 3 above)
Comment (jprider63) [reply tests/run-ci-benchmarks.sh]: Makes sense. I've added failing tests.

File: tests/run-ci-benchmarks.sh (hunk 4 above)
Comment (jprider63) [reply tests/run-ci-benchmarks.sh]: I spent a while attempting to get this to work without success. I don't think it's a proper array although I don't know bash that well. 

I'm inclined to leave the script as is for now and switch to a proper benchmarking system later.

Comment (dc-mak) [reply tests/run-ci-benchmarks.sh]: That's fine (the issue is that there needs to be an extra pair of parens on the rhs of the defining TESTS).

**Comment by jprider63** [comment]
Something is still broken with the failing tests. I'm working on debugging it.

**Comment by jprider63** [comment]
@dc-mak Are you able to reopen this or should I create a new PR? I just pushed 3a1c0bb68a696a457f1bfe7b1ee1b5fa8c31458e to this branch, which should fix the issue. I'm currently running it on our fork [here](https://github.com/GaloisInc/cerberus/actions/runs/10585515969/job/29332260659) so we'll see if that passes.

**Comment by jprider63** [comment]
FYI it [passed](https://github.com/GaloisInc/cerberus/actions/runs/10585515969/job/29332260659) on our fork.

**Comment by dc-mak** [comment]
I'm not able to do anything on this PR, so a new one will be needed.
For future PRs, please mark as draft until ready.

**Pull request merged**